packages/context/src/resolution-session.ts

```diff
@@ -181,41 +184,54 @@ export class ResolutionSession {
         ResolutionSession.describeInjection(injection),
       );
     }
-    this.injections.push(injection);
+    this.stack.push({type: 'injection', value: injection});
     /* istanbul ignore if */
     if (debugSession.enabled) {
-      debugSession('Injection path:', this.getInjectionPath());
+      debugSession('Resolution path:', this.getResolutionPath());
     }
   }
 
   /**
    * Pop the last injection
    */
   popInjection() {
-    const injection = this.injections.pop();
+    const top = this.stack.pop();
+    assert.equal(
+      top && top.type,
+      'injection',
+      'The top element must be an injection',
+    );
+    const injection = top && (top.value as Injection);
     /* istanbul ignore if */
     if (debugSession.enabled) {
       debugSession(
         'Exit injection:',
         ResolutionSession.describeInjection(injection),
       );
-      debugSession('Injection path:', this.getInjectionPath() || '<empty>');
+      debugSession('Resolution path:', this.getResolutionPath() || '<empty>');
     }
     return injection;
   }
 
   /**
    * Getter for the current injection
    */
-  get currentInjection() {
-    return this.injections[this.injections.length - 1];
+  get currentInjection(): Injection | undefined {
+    for (let i = this.stack.length - 1; i >= 0; i--) {
+      if (this.stack[i].type === 'injection')
+        return <Injection>this.stack[i].value;
```

Comment:
The necessity for explicit type cast shows a problem in the design of `ResolutionElement`, because it allows invalid combinations like `{type: 'binding', value: someInjection}`.

Have you considered using a Tagged Union Types to more accurately describe the allowed values and get better type checking from TypeScript?

```ts
export interface BindingResolution {
  type: 'binding'; 
  value: Binding; 
};

export interface InjectionResolution {
  type: 'injection'; 
  value: Injection; 
};

export type ResolutionElement = BindingResolution | InjectionResolution;
```

See https://blog.mariusschulz.com/2016/11/03/typescript-2-0-tagged-union-types and https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions